generated/summaries/data_types.md

```diff
@@ -3,18 +3,102 @@ This is the course header. This will be added on top of every page. Do to [DoDAO
+  * It is possible to query the balance of an address using the property `balance` and to send Ether (in units of wei) to a payable address using the `transfer` function.
+  * `send` is the low-level counterpart of `transfer`.
+
+- Contract types
+  * Every contract defines its own type.
+  * You can implicitly convert contracts to contracts they inherit from.
+
+- Enums
+  * Enums are used create a user-defined type in Solidity.
+  * Enums restrict the variable with one of a few predefined values, these values of the enumerated list are called enums.
+  * Syntax to declare an enum is `enum <enumerator_name> {element 1, element 2,....,element n}`
+  * Enums require at least one member, and its default value when declared is the first member.
+  * Enums cannot have more than 256 members.
+  * The options in an enum are represented by subsequent unsigned integer values starting from 0.
+
+- User Defined Value Types
+  * A user defined value type is defined using `type C is V`, where C is the name of the newly introduced type and V has to be a built-in value type (the “underlying type”). 
+  * The function C.wrap is used to convert from the underlying type to the custom type.
+  * The function C.unwrap is used to convert from the custom type to the underlying type.
+  * The type C does not have any operators or bound member functions. 
+
+- Function Types 
+  * Function types are the types of functions.
+  * Function types are declared as `function (<parameter types>) {internal|external} [pure|view|payable] [returns (<return types>)]`.
+  * Internal functions can only be called inside the current contract.
+  * External functions consist of an address and a function signature and they can be passed via and returned from external function calls.
+  * View functions are read only function and do not modify the state of the block chain (view data on the block chain).
+  * Pure functions do not read and do not modify state of the block chain.   
  
  **Reference Types**        
-Some Summary of Address   
-- Value type1
+Values of reference type can be modified through multiple different names in contrast to value types where an independent copy is maintained whenever a variable of value type is used.
+While using a reference type a data location(where the variable is stored) has to be provided explicitly. Currently, reference types comprise of structs, arrays and mappings.
+- Data location
+   * There are three data locations: memory, storage and calldata.
+   * When placed in storage, a variable is written on the blockchain. Every contract has its own storage, so these variables are persistent. 
+   * Variables stored in memory are declared inside a function. They are temporary and their lifetime is dependent on the runtime of the function they correspond to.
+   * Calldata is also a temporary data location in Solidity. It acts like memory, in terms of its dependence on the functions execution.
```

Comment:
We should expand more on calldata vs memory as its quite important topic https://ethereum.stackexchange.com/questions/74442/when-should-i-use-calldata-and-when-should-i-use-memory

Reply:
Another good article

https://medium.com/coinmonks/solidity-storage-vs-memory-vs-calldata-8c7e8c38bce